api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerInstanceTypePriceTypeAndCloudRegionTest.java

```diff
@@ -0,0 +1,707 @@
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+
+        tool = getTool(TEST_IMAGE, DEFAULT_COMMAND);
+        defaultAwsRegion = getDefaultAwsRegion(ID);
+        nonAllowedAwsRegion = getNonDefaultAwsRegion(ID_2);
+        notDefaultAwsRegion = getNonDefaultAwsRegion(ID_3);
+        configuration = getPipelineConfiguration(TEST_IMAGE, INSTANCE_DISK, true, defaultAwsRegion.getId());
+        configurationWithoutRegion = getPipelineConfiguration(TEST_IMAGE, INSTANCE_DISK);
+        price = new InstancePrice(configuration.getInstanceType(), parseInt(configuration.getInstanceDisk()),
+                PRICE_PER_HOUR, COMPUTE_PRICE_PER_HOUR, DISK_PRICE_PER_HOUR);
+        parentRun = PipelineCreatorUtils.getPipelineRunWithInstance(PARENT_RUN_ID, TEST_USER, ID_3);
+
+        mockCloudRegionManager();
+        mockInstanceOfferManager();
+        mockPreferenceManager();
+        mockPipelineConfigurationManager();
+        mockToolManager();
+
+        doReturn(parentRun).when(pipelineRunDao).loadPipelineRun(eq(PARENT_RUN_ID));
+        doReturn(false).when(permissionManager).isRunSshAllowed(eq(parentRun));
+        doReturn(DEFAULT_COMMAND).when(pipelineLauncher).launch(
+                any(PipelineRun.class), any(), any(), anyString(), anyString());
+        doReturn(ToolVersion.builder().size(1L).build())
+                .when(toolVersionManager).loadToolVersion(anyLong(), anyString());
+        doReturn(true).when(permissionHelper).isAllowed(any(), any());
+        doReturn(TEST_USER).when(securityManager).getAuthorizedUser();
+    }
+
+    @Test
+    public void shouldLaunchPipelineValidatesToolInstanceTypeAndPriceType() {
+        launchTool(configuration, INSTANCE_TYPE);
+
+        verify(toolManager, times(2)).resolveSymlinks(eq(TEST_IMAGE));
+        verify(pipelineConfigurationManager).getConfigurationForTool(eq(tool), eq(configuration));
+        verify(cloudRegionManager).load(eq(ID));
+        verify(permissionHelper).isAdmin();
+        verify(permissionHelper).isAllowed(eq(PERMISSION_NAME), eq(defaultAwsRegion));
+        verify(toolManager).loadByNameOrId(eq(TEST_IMAGE));
+
+        verify(instanceOfferManager).isToolInstanceAllowed(eq(INSTANCE_TYPE),
+                eq(getExternalResource()), eq(ID), eq(true));
+
+        verify(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), eq(getExternalResource()), eq(false));
+        verify(toolManager).getCurrentImageSize(eq(TEST_IMAGE));
+        verifyPreferenceManager(1);
+
+        verify(instanceOfferManager).getInstanceEstimatedPrice(eq(null), eq(1), eq(true), eq(ID));
+        verify(securityManager).getAuthorizedUser();
+        verify(pipelineLauncher).launch(argThat(matches(Predicates.forPipelineRun())),
+                argThat(matches(Predicates.forConfiguration())),
+                eq(null), eq("0"), eq(null));
+
+        verify(dataStorageManager).analyzePipelineRunsParameters(anyListOf(PipelineRun.class));
```

Comment:
It seems to be a little misunderstanding after the review in #1742.

I asked you to add a lot of verification to the test because the original test was testing all kind of stuff and we wanted to keep the original test behavior as much as possible. At the same time it doesn't mean that it is a good practice and definitely `FolderTemplateManagerTest` class is not a good example of a test because it has only a single god-like method that tests everything at once.

Regarding the tests in `PipelineRunManagerTest` that you moved here I would say that they are much better tests because they tested a single specification at once. If we write new or update similar tests it's a good idea to keep them in that notation: _a single test verifies a single aspect of a testing behavior_. 

In specific I would simplify this test to just the following two tests as it was originally.

```java
    @Test
    public void launchPipelineShouldValidatePriceType() {
        launchPipeline(configuration, INSTANCE_TYPE);

        verify(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), any(), anyBoolean());
    }

    @Test
    public void launchPipelineShouldValidatePipelineInstanceType() {
        launchPipeline(configuration, INSTANCE_TYPE);

        verify(instanceOfferManager).isInstanceAllowed(eq(INSTANCE_TYPE), eq(ID), eq(true));
    }
```

Could I ask to redesign tests in this class and simplify they in the same manner according to the original tests?

Reply:
All tests are significantly redesigned.